src/bson_mapper/bson_archiver.hpp

```diff
@@ -0,0 +1,1078 @@
+     */
+    void saveValue(double d) {
+        _bsonBuilder.append(d);
+    }
+
+    /**
+     * Saves a std::string to the current node.
+     *
+     * @param s
+     *    The std::string to save to the archive.
+     */
+    void saveValue(std::string const& s) {
+        _bsonBuilder.append(s);
+    }
+
+    /**
+     * Saves a c style string to the current node.
+     *
+     * @param s
+     *    The char* string to save to the archive.
+     */
+    void saveValue(char const* s) {
+        _bsonBuilder.append(s);
+    }
+
+    /**
+     * Write the name of the upcoming element and prepare object/array state.
+     * Since writeName is called for every value that is output, regardless of
+     * whether it has a name or not, it is the place where we will do a deferred
+     * check of our node state and decide whether we are in an array or an object.
+     *
+     * The general workflow of saving to the BSON archive is:
+     *   1. Set the name for the next node to be created, usually done by an NVP.
+     *   2. Start the node.
+     *   3. (if there is data to save) Write the name of the node (this function).
+     *   4. (if there is data to save) Save each element of data (with saveValue).
+     *   5. Finish the node.
+     *
+     * @param nodeInRoot
+     *  Specifies whether this is writing the name of a new node in
+     *  the root level. If this is set to true, no name will be enforced
+     *  since root level documents have no name.
+     */
+    void writeName(bool nodeInRoot = false) {
+        if (!_nodeTypeStack.empty()) {
+            OutputNodeType const& topType = _nodeTypeStack.top();
+
+            // Start up either an object or an array, depending on state.
+            if (topType == OutputNodeType::StartArray) {
+                _bsonBuilder.open_array();
+                _nodeTypeStack.top() = OutputNodeType::InArray;
+            } else if (topType == OutputNodeType::StartObject) {
+                _nodeTypeStack.top() = OutputNodeType::InObject;
+                // Only open a document if this document is not in the
+                // root node.
+                if (_nodeTypeStack.size() > 1) {
+                    _bsonBuilder.open_document();
+                }
+            }
+
+            // Elements in arrays do not have names.
+            if (topType == OutputNodeType::InArray) return;
+        }
+
+        if (!_nextName && _nodeTypeStack.empty() && nodeInRoot) {
+            // This is a document at the root node.
+            // Do nothing since no name is expected.
+            return;
+        } else if (!_nextName) {
+            // Enforce the existence of a name-value pair if this is not a node in root,
+            // or this in an element in root.
+            throw cereal::Exception("Missing a name for current node or element.\n");
+        } else {
+            // Set the key of this element to the name stored by the archiver.
+            _bsonBuilder.key_owned(_nextName);
```

Comment:
I feel like there will be cases when doing that is necessary because not every BSON key can be expressed as a C++ variable name.

I'll write a test that tries this.